### ISSUE TITLE
Allocation Unit Size

### ISSUE THREAD

**Issue by shlatchz** [dokan-dev/dokany#35]
Hi!

I tried changing:

    #define DOKAN_ALLOCATION_UNIT_SIZE	512

in public.h to 4096, and after I compiled and installed the dokan driver, the allocation size is still 512.

I tested it by creating a new txt file with just a couple of letters in it.
The size on disk appeared to be 512 bytes.

Thanks!

**Comment by Liryna** [comment]
Hi!

Well found :smiley: ! Dokan does not manage FileFsSectorSizeInformation.
https://github.com/dokan-dev/dokany/blob/22cf77c32594ec74e12038e9bc470e57524b8e9e/sys/volume.c#L73
https://msdn.microsoft.com/en-us/library/windows/hardware/ff543443%28v=vs.85%29.aspx

The size on disk 512 bytes that you see is probably the default value.

**Comment by Liryna** [comment]
@shlatchz On what OS are you testing ?
It seems that FileFsSectorSizeInformation is used only for Win8.
https://github.com/Microsoft/Windows-driver-samples/blob/3528ffbb369fa0611aefd14c97bdd6ac5ee50c41/filesys/cdfs/volinfo.c#L182

I made a test in Win7&8 and both do not care about BytesPerSector values.

**Comment by shlatchz** [comment]
I'm using Win7.
Is there a way to make it work with Win7?

**Comment by Liryna** [comment]
It is probably possible but it need more research.

**Comment by marinkobabic** [comment]
What is the output if you run 
fsutil fsinfo?

Does this hotfix make any difference
https://support.microsoft.com/en-us/kb/2618914

**Comment by shlatchz** [comment]
I've installed the hotfix.
This is the result of the command:
	fsutil fsinfo volumeinfo i:
		Volume Name : Dokan
		Volume Serial Number : 0x3a335f6
		Max Component Length : 256
		File System Name : DOKAN
		Supports Case-sensitive filenames
		Preserves Case of filenames
		Supports Unicode in filenames
		Preserves & Enforces ACL's
		Supports Remote Storage
Also, the file's size is still 512 and not 4096.

**Comment by marinkobabic** [comment]
Did you run the command elevated? Normally there should be some sector size information.

**Comment by shlatchz** [comment]
Yes I do. It won't run at all, if the cmd isn't elevated.

**Comment by marinkobabic** [comment]
What must be changed is Bytes per Cluster. As example you can run the command 

fsutil fsinfo ntfsinfo c:

against an ntfs drive. Here some details https://support.microsoft.com/en-us/kb/314878

**Comment by Liryna** [comment]
@marinkobabic, do you think we should implement FileFsSectorSizeInformation to add sector size option ?

**Comment by marinkobabic** [comment]
FileFsSizeInformation is handled in user mode code and it sets SectorsPerAllocationUnit and every system runs into this code. It should work. 

Haven't seen so far that the system executes FileFsSectorSizeInformation.

**Comment by shlatchz** [comment]
Any ideas?

**Comment by Liryna** [comment]
We should make a test with `FileFsFullSizeInformation` https://msdn.microsoft.com/en-us/library/cc232104.aspx
I just do not know when I will be able :(

**Comment by Maxhy** [comment]
I just take a look on this issue, I did some progress with commit https://github.com/dokan-dev/dokany/commit/08b09a3910dbac0e902a81ad9e9ae1a06d4a6d90 but it still doesn't work if you change DOKAN_ALLOCATION_UNIT_SIZE to another value and rebuild driver/library.
For me, it seems that correct sector/allocation unit volume information are returned if requested and then the issue doesn't lies on disk device IOCTL neither FileFsSizeInformation.
Your test is to check file size on disk after creation, which as far as I know doesn't request the volume but is based on AllocationSize field returned on file information. I tried to fix it by aligning returned AllocationSize value (changes with ALIGN_ALLOCATION_SIZE function call) ; which should have been the case btw ; but I think an implementation is missing...
